File: getshells.pl
#!/usr/bin/env perl

use strict;

my %pwhash;
my $filename = "passwd";

# Open the password file for reading
open(my $fh, '<', $filename) or die "Could not open file '$filename' $!";

# Process each line of the file
while (my $line = <$fh>) {
    chomp($line);
    # Split the line into fields using ':' as the delimiter
    my @fields = split(':', $line);
    # Extract the last field, which contains the login shell
    my $shell = $fields[-1];
    # Increment the count of accounts using this shell
    $pwhash{$shell}++;
}

# Print the output in three columns with vertical alignment
foreach my $shell (keys %pwhash) {
    # Skip empty shells
    next unless $shell;
    # Print the shell and the count of accounts using it
    printf("%-18s:\t%d\n", $shell, $pwhash{$shell});
}

# Close the file handle
close($fh);

File: getshells.rb
#!/usr/bin/env ruby

shellcnt = Hash.new(0) # use Hash.new to create a new hash with a default value of 0 for any missing keys

if File.exist?("passwd")
  File.foreach("passwd") do |line|
    line.chomp!
    gecos = line.split(':')
    shell = gecos.last
    shellcnt[shell] += 1 # use the += operator to increment the value of each shell instead of using an if/else statement
  end

  shellcnt.each do |key, value|
    printf("%-20s%-8s%-d\n", key, ":", value)
  end
end
